doc/source/ray-air/check-ingest.rst

```diff
@@ -178,6 +178,25 @@ Shuffling or data randomization is important for training high-quality models. B
      * you suspect high-quality shuffles may significantly improve model quality; and
      * absolute ingest performance is less of a concern
 
+.. _air-per-epoch-preprocessing:
+
+Applying randomized preprocessing (experimental)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The standard preprocessor passed to the ``Trainer`` is only applied once to the initial dataset when using :ref:`bulk ingest <air-streaming-ingest>`.
+However, in some cases you may want to reapply a preprocessor on each epoch, for example to augment your training dataset with a randomized transform.
+
+To support this use case, AIR offers an additional *per-epoch preprocessor* that gets reapplied on each epoch, after all other preprocessors and right before dataset consumption (e.g., using :meth:`~ray.data.DatasetIterator.iter_batches()`).
+Per-epoch preprocessing also executes in parallel with dataset consumption to reduce pauses in dataset consumption.
```

Comment:
Actually wasn't sure about this part because I don't really understand how fit() works...

- when do we need to call fit()?
- if the standard preprocessor is defined, do we need to fit() on the preprocessed dataset or the input dataset?

Reply:
Right now, we call fit on start. A fittable preprocessor isn't usable for transformation until it is fitted.

> if the standard preprocessor is defined, do we need to fit() on the preprocessed dataset or the input dataset?

Hmm I'd think you would fit the preprocessed dataset, since this preprocessor is logically consuming the output of the previous one-time preprocessor.

Perhaps we should just raise ValueError if the per-epoch preprocessor requires fitting?

Reply:
+1 on not allowing fittable per-epoch preprocessors

Reply:
This is already what we are doing here: https://github.com/ray-project/ray/pull/31725/files#diff-a3059eb0bdd6d93b67bde308d0215243a62b7de575202afda2385d9c52eb10d4R185